apps/ejabberd/src/mongoose_api_common.erl

```diff
@@ -0,0 +1,324 @@
+
+-spec compare_names_extract_args({arg_spec_list(), args_applied()}) -> arg_values().
+compare_names_extract_args({CommandsArgList, RequestArgProplist}) ->
+    Keys = lists:sort([K || {K, _V} <- RequestArgProplist]),
+    ExpectedKeys = lists:sort([Key || {Key, _Type} <- CommandsArgList]),
+    ZippedKeys = lists:zip(Keys, ExpectedKeys),
+    case lists:member(false, [ReqKey =:= ExpKey || {ReqKey, ExpKey} <- ZippedKeys]) of
+        true ->
+            throw({error, bad_request, ?ARGS_SPEC_ERROR});
+        _ ->
+            do_extract_args(CommandsArgList, RequestArgProplist)
+    end.
+
+-spec do_extract_args(arg_spec_list(), args_applied()) -> arg_values().
+do_extract_args(CommandsArgList, RequestArgList) ->
+    [element(2, lists:keyfind(Key, 1, RequestArgList)) || {Key, _Type} <- CommandsArgList].
+
+-spec execute_command(list({atom(), any()}) | map() | {error, atom(), any()}, mongoose_command(), admin|binary()) ->
+                     {ok, term()} | failure().
+execute_command({error, _Type, _Reason} = Err, _, _) ->
+    Err;
+execute_command(Args, Command, Entity) ->
+    try
+        do_execute_command(Args, Command, Entity)
+    catch
+        _:R ->
+            {error, bad_request, R}
+    end.
+
+-spec do_execute_command(arg_values(), mongoose_command(), admin|binary()) -> ok | {ok, any()}.
+do_execute_command(Args, Command, Entity) ->
+    Types = [Type || {_Name, Type} <- ?COMMANDS_ENGINE:args(Command)],
+    ConvertedArgs = [convert_arg(Type, Arg) || {Type, Arg} <- lists:zip(Types, Args)],
+    ?COMMANDS_ENGINE:execute(Entity, ?COMMANDS_ENGINE:name(Command), ConvertedArgs).
+
+-spec maybe_add_caller(admin | binary) -> list() | list({caller, binary()}).
+maybe_add_caller(admin) ->
+    [];
+maybe_add_caller(JID) ->
+    [{caller, JID}].
+
+-spec maybe_add_location_header(integer() | binary() | list() | float(), list(), any()) -> any().
+maybe_add_location_header(Result, ResourcePath, Req) when is_binary(Result) ->
+    add_location_header(binary_to_list(Result), ResourcePath, Req);
+maybe_add_location_header(Result, ResourcePath, Req) when is_list(Result) ->
+    add_location_header(Result, ResourcePath, Req);
+maybe_add_location_header(Result, ResourcePath, Req) when is_integer(Result) ->
+    add_location_header(integer_to_list(Result), ResourcePath, Req);
+maybe_add_location_header(Result, ResourcePath, Req) when is_float(Result) ->
+    add_location_header(float_to_list(Result), ResourcePath, Req);
+maybe_add_location_header(_R, _R, Req) ->
+    {ok, Req2} = cowboy_req:reply(200, [], Req),
+    Req2.
+
+add_location_header(Result, ResourcePath, Req) ->
+    Path = ResourcePath ++ "/" ++ Result,
+    Header = {<<"location">>, Path},
+    {ok, Req2} = cowboy_req:reply(201, [Header], Req),
+    Req2.
+
+
+
+-spec convert_arg(atom(), any()) -> integer() | float() | binary() | string() | {error, bad_type}.
+convert_arg(binary, Binary) when is_binary(Binary) ->
+    Binary;
+convert_arg(string, Binary) when is_binary(Binary) ->
```

Comment:
Why do we need `string` type? Can `binary` hold sth else than `string` can in the API case?
If no, I think we should have only `binary` type (mainly for simplicity).

Please correct me if I'm missing sth.